src/Language/PureScript/TypeChecker.hs

```diff
@@ -584,13 +584,17 @@ typeCheckAll moduleName = traverse go
       , ..
       }
 
+-- | Check that a newtype has just one data constructor with just one field, or
+-- throw an error. If the newtype is valid, this function returns the single
+-- data constructor declaration and the single field, as a 'proof' that the
+-- newtype was indeed a valid newtype.
```

Comment:
Returning the decl and the field here means that we avoid an unsafe pattern match in `Sugar/TypeClasses/Deriving`.